File: day6.c
#include <stdio.h>

typedef struct {
    long long int time;
    long long int distance;
} Race;

long long int wayWin(long long int time, long long int dist) {
    long long int ways = 0;
    for (long long int hold_time = 0; hold_time < time; ++hold_time) {
        long long int mt = time - hold_time;
        long long total_distance = (long long)mt * hold_time;
        if (total_distance > dist) {
            ways++;
        }
    }
    return ways;
}

int main() {
    Race races[] = {
        {49, 298},
        {78, 1185},
        {79, 1066},
        {80, 1181}
    };
    long long int num_races = sizeof(races) / sizeof(races[0]);

    long long total = 1;
    for (long long int i = 0; i < num_races; ++i) {
        total *= wayWin(races[i].time, races[i].distance);
    }
    printf("Part 1: %lld\n", total);

    long long int single_time = 49787980;
    long long single_dist = 298118510661181LL;
    printf("Part 2: %lld\n", wayWin(single_time, single_dist));

    return 0;
}

File: day6.py
INP = """Time:        49     78     79     80
Distance:   298   1185   1066   1181"""

races = [
    {'time': 49, 'distance': 298},
    {'time': 78, 'distance': 1185},
    {'time': 79, 'distance': 1066},
    {'time': 80, 'distance': 1181}
]

def wayWin(time, dist):
    ways = 0
    for hold_time in range(time):
        mt = time - hold_time
        total_distance = mt * hold_time
        if total_distance > dist:
            ways += 1
    return ways

ways_p_race = [wayWin(race["time"], race["distance"]) for race in races]

total = 1
for ways in ways_p_race: total *= ways

print("Part  1:", total)
single_time = 49787980
single_dist = 298118510661181
print("Part  2:", wayWin(single_time, single_dist))